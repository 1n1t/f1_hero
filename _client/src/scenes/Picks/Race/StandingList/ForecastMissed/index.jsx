import React from 'react';

const ForecastMissed = () => (
	<div className="standings-container">
		<div className="standing">
			You did not submit your prediction on time. No race results available for
			you.
		</div>
	</div>
);

export default ForecastMissed;
