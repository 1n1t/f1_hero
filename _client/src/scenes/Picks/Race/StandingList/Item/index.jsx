import React from 'react';
import PropTypes from 'prop-types';

import './styles.scss';

const Item = ({ score, position, _driver }) => {
	const icon =
		score !== null ? (
			<div className="score">{score}pt</div>
		) : (
			<i className="fa fa-sort" aria-hidden="true" />
		);

	return (
		<div className="StandingItem">
			<div className="position">{position}.</div>
			<div className="driver-name">{_driver.name}</div>
			<div className="driver-code">{_driver.code}</div>
			{icon}
		</div>
	);
};

Item.defaultProps = {
	score: null
};

Item.propTypes = {
	score: PropTypes.number,
	position: PropTypes.number.isRequired,
	_driver: PropTypes.shape({
		name: PropTypes.string.isRequired,
		code: PropTypes.string.isRequired
	}).isRequired
};

export default Item;
