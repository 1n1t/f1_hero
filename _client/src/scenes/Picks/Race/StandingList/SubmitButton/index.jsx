import React from 'react';
import PropTypes from 'prop-types';

import './styles.scss';

const SubmitButton = ({ onClick }) => (
	<div className="btn SubmitButton" onClick={onClick} role="button" tabIndex="0">
		<i className="fa fa-check" aria-hidden="true" /> Submit
	</div>
);

SubmitButton.propTypes = {
	onClick: PropTypes.func.isRequired
};

export default SubmitButton;
