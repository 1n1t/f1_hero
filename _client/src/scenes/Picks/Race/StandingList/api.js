/* eslint-disable import/prefer-default-export */
import fetchApi from '../../../../services/api';

const endPoints = {
	postPick(round) {
		return `/api/picks/${round}`;
	}
};

export const postPick = (round, payload) =>
	fetchApi(endPoints.postPick(round), 'post', payload);
