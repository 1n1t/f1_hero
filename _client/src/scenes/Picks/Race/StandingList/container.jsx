/* eslint-disable react/prop-types */
import React from 'react';
import { connect } from 'react-redux';

import StandingList from './index';
import { updateRace } from '../../../../data/races/actions';

const StandingListContainer = props => <StandingList {...props} />;

export const mapStateToProps = state => ({
	races: state.data.races.all,
	drivers: state.data.drivers.all
});

export default connect(mapStateToProps, { updateRace })(StandingListContainer);
