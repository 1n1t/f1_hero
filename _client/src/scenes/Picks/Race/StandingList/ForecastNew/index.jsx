import React from 'react';
import PropTypes from 'prop-types';
import { SortableContainer, SortableElement } from 'react-sortable-hoc';

import Item from '../Item';
import { listItemPropTypes } from '../index';

const ForecastNew = ({ items }) => {
	const SortableItem = SortableElement(item => <Item {...item} />);

	return (
		<div className="standings-container">
			{items.map((value, index) => (
				<SortableItem key={`item-${value.position}`} index={index} {...value} />
			))}
		</div>
	);
};

ForecastNew.propTypes = {
	items: PropTypes.arrayOf(PropTypes.shape(listItemPropTypes)).isRequired
};

export default SortableContainer(ForecastNew);
