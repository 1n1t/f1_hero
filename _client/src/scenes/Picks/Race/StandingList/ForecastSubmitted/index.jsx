import React from 'react';
import PropTypes from 'prop-types';

import Item from '../Item';
import { listItemPropTypes } from '../index';

const ForecastSubmitted = ({ items }) => (
	<div className="standings-container">
		{items.map(item => <Item key={item.position} {...item} />)}
	</div>
);

ForecastSubmitted.propTypes = {
	items: PropTypes.arrayOf(PropTypes.shape(listItemPropTypes)).isRequired
};

export default ForecastSubmitted;
