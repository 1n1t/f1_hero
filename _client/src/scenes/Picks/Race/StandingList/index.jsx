/* eslint-disable jsx-a11y/click-events-have-key-events */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { arrayMove } from 'react-sortable-hoc';
import { calculatePickAndStemLeftPosition } from 'App/services/utils/functions';

import ForecastNew from './ForecastNew';
import ForecastMissed from './ForecastMissed';
import ForecastSubmitted from './ForecastSubmitted';
import SubmitButton from './SubmitButton';
import { postPick } from './api';
import { racePropType } from '../index';
import './styles.scss';

class StandingList extends Component {
	state = {
		pickPos: 0,
		stemPos: 0,
		list:
			this.props.status === 'submitted' ? this.props.list : this.props.drivers,
		status: this.props.status
	};

	componentDidMount() {
		const { races, round, raceNode, raceHolderNode } = this.props;
		const { pickPos, stemPos } = calculatePickAndStemLeftPosition(
			{
				picksNode: this.picks,
				stemNode: this.stem,
				raceNode,
				raceHolderNode
			},
			races.length,
			round
		);

		const newState = {
			...this.state,
			pickPos,
			stemPos
		};
		this.setState(newState);
	}

	onSortEnd = ({ oldIndex, newIndex }) => {
		const newList = arrayMove(this.state.list, oldIndex, newIndex);
		const formattedList = newList.map((value, index) => ({
			...value,
			position: index + 1
		}));

		const newState = {
			...this.state,
			list: formattedList //	todo make list array immutable
		};
		this.setState(newState);
	};

	submitPick = async () => {
		const { round, onSubmit, updateRace } = this.props;
		try {
			const { pick } = (await postPick(round, {
				round,
				forecast: this.state.list.slice(0, 10)
			})).data;

			const newState = {
				...this.state,
				status: 'submitted',
				list: pick.forecast
			};

			this.setState(newState);
			onSubmit(pick);
			updateRace({ round, field: 'hasPick', value: true });
		} catch (error) {
			//	todo display an error as a flash message
			console.error('error: unable to save pick!!', error);
		}
	};

	render() {
		const { status } = this.state;
		let Standings;

		switch (status) {
			case 'new':
				Standings = ForecastNew;
				break;
			case 'missed':
				Standings = ForecastMissed;
				break;
			default:
				Standings = ForecastSubmitted;
		}

		return (
			<div
				ref={el => {
					this.picks = el;
				}}
				className="StandingList"
				style={{ left: `${this.state.pickPos}px` }}
			>
				<div
					ref={el => {
						this.stem = el;
					}}
					className="stem"
					style={{ left: `${this.state.stemPos}px` }}
				/>
				<Standings
					items={this.state.list}
					onSortEnd={this.onSortEnd}
					lockToContainerEdges
					helperClass="draggable"
					lockAxis="y"
				/>
				{status === 'new' && <SubmitButton onClick={this.submitPick} />}
			</div>
			//	todo make function that will create refs
		);
	}
}

export const listItemPropTypes = {
	position: PropTypes.number.isRequired,
	_drivers: PropTypes.arrayOf(
		PropTypes.shape({
			code: PropTypes.string.isRequired,
			name: PropTypes.string.isRequired
		})
	)
};

StandingList.defaultProps = {
	list: [],
	drivers: [],
	onSubmit: () => {}
};

StandingList.propTypes = {
	status: PropTypes.oneOf(['new', 'submitted', 'passed']).isRequired,
	list: PropTypes.arrayOf(PropTypes.shape(listItemPropTypes)),
	drivers: PropTypes.arrayOf(PropTypes.shape(listItemPropTypes)),
	round: PropTypes.number.isRequired,
	races: PropTypes.arrayOf(PropTypes.shape(racePropType)).isRequired,
	onSubmit: PropTypes.func,
	updateRace: PropTypes.func.isRequired,
	// eslint-disable-next-line react/forbid-prop-types
	raceHolderNode: PropTypes.object.isRequired,
	// eslint-disable-next-line react/forbid-prop-types
	raceNode: PropTypes.object.isRequired
};

export default StandingList;
