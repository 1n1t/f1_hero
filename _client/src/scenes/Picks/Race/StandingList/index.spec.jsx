import r from 'Root/jest/utils/renderer';
import pickSample from 'JestClient/samples/pick';
import driversSample from 'JestClient/samples/drivers';
import racesSample from 'JestClient/samples/races';
import moxios from 'moxios';

import StandingList from './index';
import ForecastNew from './ForecastNew';
import ForecastMissed from './ForecastMissed';
import ForecastSubmitted from './ForecastSubmitted';
import Item from './Item';
import SubmitButton from './SubmitButton';
import { endPoints } from './api';

const defaultProps = {
	status: 'submitted',
	list: pickSample.forecast,
	drivers: driversSample.all,
	round: 1,
	races: racesSample.all,
	onSubmit() {},
	updateRace() {}
};

r.init(StandingList, defaultProps);

it('renders correctly', () => {
	const wrapper = r.render();
	expect(wrapper).toMatchSnapshot();
});

describe('renders correct forecast component', () => {
	it('when provided status is new', () => {
		const wrapper = r.render({ status: 'new' });

		expect(wrapper.find(ForecastNew)).toHaveLength(1);
		expect(wrapper.find(ForecastMissed)).toHaveLength(0);
		expect(wrapper.find(ForecastSubmitted)).toHaveLength(0);
	});

	it('when provided status is missed', () => {
		const wrapper = r.render({ status: 'missed' });

		expect(wrapper.find(ForecastMissed)).toHaveLength(1);
		expect(wrapper.find(ForecastNew)).toHaveLength(0);
		expect(wrapper.find(ForecastSubmitted)).toHaveLength(0);
	});

	it('when provided status is submitted', () => {
		const wrapper = r.render({ status: 'submitted' });

		expect(wrapper.find(ForecastSubmitted)).toHaveLength(1);
		expect(wrapper.find(ForecastNew)).toHaveLength(0);
		expect(wrapper.find(ForecastMissed)).toHaveLength(0);
	});
});

describe('submit button', () => {
	it('is shown when user is picking a new race forecast', () => {
		const wrapper = r.render({ status: 'new' });
		const button = wrapper.find(SubmitButton);

		expect(button).toHaveLength(1);
	});

	it('is hidden when user is checking previously submitted forecast', () => {
		const wrapper = r.render({ status: 'submitted' });
		const button = wrapper.find(SubmitButton);

		expect(button).toHaveLength(0);
	});

	it('is hidden when user is checking previously missed forecast', () => {
		const wrapper = r.render({ status: 'missed' });
		const button = wrapper.find(SubmitButton);

		expect(button).toHaveLength(0);
	});
});

it("submits user's forecast", done => {
	const onSubmit = jest.fn();
	const updateRace = jest.fn();
	const wrapper = r.render({ status: 'new', onSubmit, updateRace });
	const button = wrapper.find(SubmitButton);

	moxios.withMock(() => {
		button.simulate('click');

		moxios.wait(async () => {
			const request = moxios.requests.mostRecent();
			const response = { pick: pickSample };

			await request.respondWith({
				status: 200,
				response
			});

			expect(request.config.method).toBe('post');
			expect(request.url).toBe(endPoints.postPick(defaultProps.round));

			const standingsEl = wrapper.update().find(ForecastSubmitted);
			expect(standingsEl).toHaveLength(1);

			const firstListItem = standingsEl
				.dive()
				.find(Item)
				.at(0);
			expect(firstListItem.props()._driver.name).toBe(
				response.pick.forecast[0]._driver.name
			);

			expect(onSubmit).toHaveBeenCalledTimes(1);
			expect(onSubmit).toHaveBeenCalledWith(response.pick);

			expect(updateRace).toHaveBeenCalledTimes(1);
			expect(updateRace).toHaveBeenCalledWith({
				round: defaultProps.round,
				field: 'hasPick',
				value: true
			});

			done();
		});
	});
});

it('can sort new forecast list', () => {
	const oldIndex = 1;
	const newIndex = 0;
	const wrapper = r.render({ status: 'new' });
	const initialList = wrapper.state('list');

	const forecast = wrapper.find(ForecastNew);
	forecast.props().onSortEnd({ oldIndex, newIndex });

	const newList = wrapper.update().state('list');

	expect(newList[newIndex]._driver).toMatchObject(
		initialList[oldIndex]._driver
	);
	expect(newList[oldIndex]._driver).toMatchObject(
		initialList[newIndex]._driver
	);

	expect(newList[0].position).toBe(1);
	expect(newList[1].position).toBe(2);
});
