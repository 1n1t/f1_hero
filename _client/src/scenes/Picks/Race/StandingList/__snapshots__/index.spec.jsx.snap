// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="StandingList"
  style={
    Object {
      "left": "0px",
    }
  }
>
  <div
    className="stem"
    style={
      Object {
        "left": "0px",
      }
    }
  />
  <ForecastSubmitted
    helperClass="draggable"
    items={
      Array [
        Object {
          "_driver": Object {
            "code": "HAM",
            "name": "L. Hamilton",
          },
          "position": 1,
          "score": 18,
        },
        Object {
          "_driver": Object {
            "code": "RIC",
            "name": "D. Ricciardo",
          },
          "position": 2,
          "score": 5,
        },
        Object {
          "_driver": Object {
            "code": "VAN",
            "name": "S. Vandoorne",
          },
          "position": 3,
          "score": 0,
        },
      ]
    }
    lockAxis="y"
    lockToContainerEdges={true}
    onSortEnd={[Function]}
  />
</div>
`;
